juju/terraform-provider-juju
# Use juju 3 5 1

## Description

Fixing the issue reported in #475 by pulling in the newer juju api versions. Will be validated by @babakks during RC.

Testing the provider with juju 3.5.1 APIs, found a bug in our provider configuration code where juju login details are discovered. Both client ID/secret and username/password were being provided to the juju client. These represent the 2 methods for juju login possible, however only 1 may be used at any time. The former method is for JIMM and the latter is used for directly juju login. Client ID/Secret can be found either in the plan and/or specific environment variables. Username/password can be found the same way and via discovery by using the juju cli. 

Use the terraform plan is the principal definition of the juju login details for the provider. If pieces are missing, attempt to find them from the specific environment variables. At this point if the Client ID/Secret method is being used, all pieces are required, or we error. If username/password are being used and all pieces are not available, attempt discovery via the juju cli. 

Directly use only 1 version of a package, in this case the juju/charm package.

## Type of change

- Other (please describe)

## QA steps

All current integration tests are successful.

Scenarios depend on specific controller used for test. The test scenarios below will contain a framework for tests, you must add your controller details. It's recommended to put the certificate in a file.

All potential information for juju login inside of a plan is listed below. Using all should fail with a message indicating the problem. 
```
provider "juju" {
  controller_addresses = <controller-ip-addrs>
  username = "admin"
  password = <password>
  ca_certificate = file("./controller.pem")
  client_id = "testme"
  client_secret = "secret"
}
```

Possible environment variables are as follows:
```
JUJU_CONTROLLER_ADDRESSES
JUJU_USERNAME
JUJU_PASSWORD
JUJU_CA_CERT
JUJU_CLIENT_ID
JUJU_CLIENT_SECRET
```

Unless you have a JIMM config, using the client ID/secret method won't succeed. However it'll be an juju error message, not directly from the terraform provider.

Please test combinations of the 3 methods to provide juju login for configuration within the limited described. An informative error message should appear when the correct details are not available.

## Additional notes

https://warthogs.atlassian.net/browse/JUJU-6156
